src/cai_lab_to_nwb/zaki_2024/zaki_2024_convert_week_session.py

```diff
@@ -65,28 +72,83 @@ def session_to_nwb(
+            identifier=str(uuid.uuid4()),
+            session_start_time=session_start_time,
+        )
+    )
+
+    nwbfile = NWBFile(**nwbfile_kwargs)
+
+    # Add epochs table to store time range of conditioning and offline sessions
+    sessions_summary_file = data_dir_path / f"Ca_EEG_Experiment/{subject_id}/{subject_id}_SessionTimes.csv"
+    sessions_summary_df = pd.read_csv(sessions_summary_file)
+
+    # Add columns to TimeIntervals
+    nwbfile.add_epoch_column(name="session_ids", description="ID of the session")
+
+    for task, date_str, time_str in zip(
+        sessions_summary_df["Session"], sessions_summary_df["Date"], sessions_summary_df["Time"]
+    ):
+        session_id = subject_id + "_" + task
+        if "Offline" in session_id:
+            offline_day = session_id.split("Session")[0]
+            experiment_dir_path = (
+                data_dir_path / "Ca_EEG_Experiment" / subject_id / (subject_id + "_Offline") / offline_day
+            )
+        else:
+            experiment_dir_path = (
+                data_dir_path / "Ca_EEG_Experiment" / subject_id / (subject_id + "_Sessions") / session_id
+            )
+        try:
+            folder_path = experiment_dir_path / date_str / time_str
+            miniscope_folder_path = get_miniscope_folder_path(folder_path)
+            miniscope_metadata_json = folder_path / "metaData.json"
+            assert miniscope_metadata_json.exists(), f"General metadata json not found in {folder_path}"
+            timestamps_file_path = miniscope_folder_path / "timeStamps.csv"
+            assert timestamps_file_path.exists(), f"Miniscope timestamps file not found in {miniscope_folder_path}"
+
+            start_datetime_timestamp, stop_datetime_timestamp = get_session_slicing_time_range(
+                miniscope_metadata_json=miniscope_metadata_json, timestamps_file_path=timestamps_file_path
+            )
+
+            start_time = (start_datetime_timestamp - session_start_time.replace(tzinfo=None)).total_seconds()
+            stop_time = (stop_datetime_timestamp - session_start_time.replace(tzinfo=None)).total_seconds()
+
+        # Some sessions may not have imaging data, so we extract the run time from the session notes (.txt file)
+        # and use the data string and time string to retrieve the start datetime of the session
+        except:
+            datetime_str = date_str + " " + time_str
+            start_datetime_timestamp = datetime.strptime(datetime_str, "%Y_%m_%d %H_%M_%S")
+
+            txt_file_path = experiment_dir_path / f"{session_id}.txt"
+            session_run_time = get_session_run_time(txt_file_path=txt_file_path)
+
+            start_time = (start_datetime_timestamp - session_start_time.replace(tzinfo=None)).total_seconds()
+            stop_time = start_time + session_run_time
+
+        nwbfile.add_epoch(start_time=start_time, stop_time=stop_time, session_ids=session_id)
+
+    converter = Zaki2024NWBConverter(source_data=source_data)
```

Comment:
Another pattern that you can use here is to first create the converter, then create the nwbfile (`converter.create_nwbfile()`) then you add the epochs to the created nwbfile and write it.

Reply:
Right! Thanks, I'll try this